File: 2021_04_18_make_a_lisp_in_csharp/Mal/Mal.Tests/TypesTests.cs

using System;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using PowerFP;
using static Mal.Types;


namespace Mal.Tests
{
    [TestClass]
    public class TypesTests
    {
        [TestMethod]
        public void MalEqualTest()
        {
            Assert.IsTrue(MalEqual(new Str("a"), new Str("a")));
            Assert.IsFalse(MalEqual(new Str("a"), new Str("b")));
            Assert.IsFalse(MalEqual(new Str("a"), new Symbol("b", NilV)));

            Func<LList<MalType>?, MalType> f = _ => NilV;
            Assert.IsTrue(MalEqual(new Fn(f, NilV), new Fn(f, NilV)));
            Assert.IsFalse(MalEqual(new Fn(_ => NilV, NilV), new Fn(f, NilV)));


            Assert.IsTrue(MalEqual(
                new Map(MapM.MapFrom<MalType, MalType>((new Str("name"), new Str("marcin"))), NilV),
                new Map(MapM.MapFrom<MalType, MalType>((new Str("name"), new Str("marcin"))), NilV)));

            Assert.IsFalse(MalEqual(
                new Map(MapM.MapFrom<MalType, MalType>((new Keyword("name"), new Str("marcin"))), NilV),
                new Map(MapM.MapFrom<MalType, MalType>((new Str("name"), new Str("marcin"))), NilV)));


            Assert.IsFalse(MalEqual(
                new Map(MapM.MapFrom<MalType, MalType>((new Str("name"), new Keyword("marcin"))), NilV),
                new Map(MapM.MapFrom<MalType, MalType>((new Str("name"), new Str("marcin"))), NilV)));

            Assert.IsFalse(MalEqual(
                new Map(MapM.MapFrom<MalType, MalType>((new Str("name"), new Keyword("marcin")), (new Str("age"), new Number(30))), NilV),
                new Map(MapM.MapFrom<MalType, MalType>((new Str("name"), new Str("marcin"))), NilV)));
        }
    }
}

File: 2021_10_21_make_a_lisp_in_fsharp/MiniMal/MiniMal.Tests/TypesTests.fs
namespace MiniMal.Tests

open Microsoft.VisualStudio.TestTools.UnitTesting
open Types
open TestUtils

[<TestClass>]
type TypesTests() =

    [<TestMethod>]
    member this.MalEqualTest() =
        let fn = Fn((fun args -> Nil), false)
        assertMalNotEquals fn fn

        assertMalEquals False False
        assertMalNotEquals False fn

        assertMalEquals True True
        assertMalNotEquals True fn

        assertMalEquals Nil Nil
        assertMalNotEquals Nil fn

        assertMalEquals (Str("a")) (Str("a"))
        assertMalNotEquals (Str("a")) (Str("b"))
        assertMalNotEquals (Str("a")) fn

        assertMalEquals (Number(1.)) (Number(1.))
        assertMalNotEquals (Number(1.)) (Number(2.))
        assertMalNotEquals (Number(1.)) fn

        assertMalEquals (Symbol("a")) (Symbol("a"))
        assertMalNotEquals (Symbol("a")) (Symbol("b"))
        assertMalNotEquals (Symbol("a")) fn

        let mals = [ Str("name"); Nil ]
        assertMalEquals (MalList(mals, ListType.List)) (MalList(mals |> (List.map id), ListType.List))
        assertMalEquals (MalList(mals, ListType.List)) (MalList(mals |> (List.map id), ListType.Vector))
        assertMalNotEquals (MalList(mals, ListType.List)) (MalList(mals |> (List.take 1), ListType.List))
        assertMalNotEquals (MalList(mals, ListType.List)) (MalList(mals |> List.rev, ListType.List))
        assertMalNotEquals (MalList(mals, ListType.List)) fn

        let map = Map [ "name", Nil ]
        let malMap = MalMap(map)
        assertMalEquals malMap (MalMap(map |> Map.toList |> (List.map id) |> Map.ofList))

        assertMalNotEquals
            malMap
            (MalMap(
                map
                |> Map.toList
                |> (List.map (fun (k, v) -> (k + "!", v)))
                |> Map.ofList
            ))


        assertMalNotEquals malMap (MalMap(Map []))
        assertMalNotEquals malMap fn

        let map1 =
            Map([ ("name", Nil); ("age", Number(1.)) ])
        let map2 =
            Map([ ("age", Number(1.)); ("name", Nil) ])
        assertMalEquals (MalMap(map1)) (MalMap(map2))
